tools/GenerateXdmf.py

```diff
@@ -47,52 +47,100 @@ def generate_xdmf(file_prefix, output_filename, start_time, stop_time, stride):
         for h5file in h5files:
             h5temporal = h5file[0].get('element_data.vol').get(id_and_value[0])
+            dimensionality = 3
+            # If there are no z-coordinates then assume the data is 2d. If
+            # in the future this assumption is invalid on datasets we will
+            # need an extra command line argument.
+            if not 'InertialCoordinates_z' in list(h5temporal.keys()):
+                dimensionality = 2
             extents = h5temporal.get("total_extents")
-            extents_x = np.array([extents[i] for i in
-                                  range (extents.size) if i%3 == 0])
-            extents_y = np.array( [extents[i] for i in
-                                   range (extents.size) if i%3 == 1])
-            extents_z = np.array( [extents[i] for i in
-                                   range (extents.size) if i%3 == 2])
-
-            total_extents_x = sum(extents_x)
-            total_extents_y = sum(extents_y)
-            total_extents_z = sum(extents_z)
-            numpoints = sum(extents_x*extents_y*extents_z)
-            number_of_cells = sum((extents_x-1)*(extents_y-1)*(extents_z-1))
+            extents_x = np.array([
+                extents[i] for i in range(extents.size)
+                if i % dimensionality == 0
+            ])
+            extents_y = np.array([
+                extents[i] for i in range(extents.size)
+                if i % dimensionality == 1
+            ])
+
+            # So that we can have more generic code, set that we have 1
+            # grid point in z in 2d
+            extents_z = np.array([1])
+            if dimensionality == 3:
+                extents_z = np.array([
+                    extents[i] for i in range(extents.size)
+                    if i % dimensionality == 2
+                ])
+
+            numpoints = sum(extents_x * extents_y * extents_z)
+            # We can't have zero cells if rendering a 2d surface, so
+            # return 1 in that case by using a ternary.
+            number_of_cells = sum(
+                (extents_x - 1) * (extents_y - 1) *
+                (extents_z - 1 if dimensionality == 3 else 1))
+
             data_item = "        <DataItem Dimensions=\" %d\" " \
                     "NumberType=\"Double\" Precision=\"8\" Format=\"HDF5\">\n" \
                     % (numpoints)
-            data_item_vec = "        <DataItem Dimensions=\" %d 3\" "\
+
+            # Set up vectors
+            if dimensionality == 3:
+                data_item_vec = "        <DataItem Dimensions=\" %d 3\" "\
                     "ItemType = \"Function\" Function = \"JOIN($0,$1,$2)\">\n"\
                             % (numpoints)
-            Grid_path = "          %s:/element_data.vol/%s" % (
-                h5file[1], id_and_value[0])
+            else:
+                # In 2d we still need a 3d dataset to have a vector because
+                # ParaView only supports 3d vectors. We deal with this by making
```

Comment:
Could it be useful to allow full 3d vectors on 2 dimensional domains?  Not sure if this is feasible but it might be useful to allow vectors "coming out of the plane" when visualizing data.